presto-main/src/main/java/com/facebook/presto/sql/planner/SqlPlannerContext.java

```diff
@@ -64,57 +63,22 @@ public void incrementLeafNodes(Session session)
         }
     }
 
-    public class NestedCteStack
+    public class CteInfo
     {
         @VisibleForTesting
         public static final String delimiter = "_*%$_";
-        private final Stack<String> cteStack;
-        private final Map<String, String> rawCtePathMap;
+        // never decreases
+        private int currentQueryScopeId;
 
-        public NestedCteStack()
-        {
-            this.cteStack = new Stack<>();
-            this.rawCtePathMap = new HashMap<>();
-        }
-
-        public void push(String cteName, Query query)
-        {
-            this.cteStack.push(cteName);
-            if (query.getWith().isPresent()) {
-                // All ctes defined in this context should have their paths updated
-                query.getWith().get().getQueries().forEach(with -> this.addNestedCte(with.getName().toString()));
-            }
-        }
-
-        public void pop(Query query)
-        {
-            this.cteStack.pop();
-            if (query.getWith().isPresent()) {
-                query.getWith().get().getQueries().forEach(with -> this.removeNestedCte(with.getName().toString()));
-            }
-        }
+        // All references of the same queryStatement fall under the same scope
+        Map<Query, String> queryNodeScopeIdMap = new HashMap<>();
```

Comment:
nit: why do we convert the scope id to string?

Reply:
Just so that it is easier to debug and add tests, we can use hashCode but it becomes complicated